docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+
+```
+Commands
+  Develop an app
+    init     	: Initialize a new application.
+    provision	: Provision the Azure resources for an application.
+    deploy   	: Deploy the application's code to Azure.
+    up       	: Initialize application, provision Azure resources, and deploy your project with a single command.
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
+    restore  	: Restore application dependencies.
+    template 	: Find and view template details.
+```
+
+The commands are grouped by their scenarios, with most important scenarios at the top. It is not in alphabetical order, rather the order the user is expected to run the commands. If we added `package` for example, would be after `init`.
+
+## Command changes
+
+The following changes are proposed:
+
+- `azd infra create` and `azd infra delete` should be removed, as the functionality already exists in `azd provision` and `azd down`, and users will find the duplication somewhat confusing.
+- The `--output` flag should be moved to `experimental` (hidden and unsupported unless with feature toggle), with the exception of `azd auth token --output json` which must be supported due to credential handlers.
+  - When the feature is hidden, the default output behavior of each command will apply.
+  - This is done as `--output` does not have a documented contract. It prints out `stdout` in JSON, while `stderr` messages also contain JSON-enveloped messages. We might consider splitting this behavior in the future.
+
+Discussion warranted:
+
+- Should we offer `azd restore`? I'm curious if users actually find this useful. We currently use this in our `tasks.json`, which seems useful, but the immediate block directly references `npm`, which defeats the nice abstraction. Excerpt from `tasks.json` below:
+  ```json
+    {
+    "label": "Restore Web",
+    "type": "shell",
+    "command": "azd restore --service web",
+    "presentation": {
+        "reveal": "silent"
+    },
+    "problemMatcher": []
+    },
+    {
+        "label": "Web npm start",
+        "detail": "Helper task--use 'Start Web' task to ensure environment is set up correctly",
+        "type": "shell",
+        "command": "npm run start"
+    }
+  ```
+- Should `azd deploy` deploy only the current service, when the working directory is under a service directory? I personally think this is what users might expect, but I can see an argument against it since it may be implicit behavior.
```

Comment:
I think we should just default to the explicit `--service` flag to avoid scenarios where people run into unexpected behaviour.

Reply:
We still have to decide on the behavior of running under a service folder. Not sure what happens today but you could say that it doesn't do anything because there is no azure.yaml. You could have it walk up the folders to find an azure.yaml and do all the services, or you could have it walk up the folders to find an azure.yaml and do just the one from the folder you are in. Have we spec'd out our probing behavior for looking up the folder structure for an azure.yaml?

Reply:
We already do this currently. `azure.yaml` is looked up recursively going up the directories from CWD until one is found. If none, `ErrNoProject` is returned.

The proposal merely extends it to also match against the defined `services` in `azure.yaml` before deciding what to invoke.